state-chain/runtime/src/chainflip.rs

```diff
@@ -341,16 +345,42 @@ impl TransactionBuilder<Solana, SolanaApi<SolEnvironment>> for SolanaTransaction
 	}
 
 	fn requires_signature_refresh(
-		_call: &SolanaApi<SolEnvironment>,
+		call: &SolanaApi<SolEnvironment>,
 		_payload: &<<Solana as Chain>::ChainCrypto as ChainCrypto>::Payload,
-		_maybe_current_onchain_key: Option<<SolanaCrypto as ChainCrypto>::AggKey>,
-	) -> bool {
-		// We use Durable Nonce mechanism to avoid the 150 blocks expiry period and so
-		// transactions won't expire. Then, the only reason to resign would be if the
-		// payload has been updated or the aggKey has been updated (key rotation).
-		// The payload won't be refreshed, as explained above, and the the broadcast
-		// barrier prevents a transaction from requiring to be resigned by a new aggKey.
-		false
+		maybe_current_onchain_key: Option<<SolanaCrypto as ChainCrypto>::AggKey>,
+	) -> RequiresSignatureRefresh<SolanaCrypto, SolanaApi<SolEnvironment>> {
+		// The only reason to resign would be if the aggKey has been updated on chain (key rotation)
+		// and this apicall is signed with the old key and is still pending. In this case, we need
+		// to modify the apicall, by replacing the aggkey with the new key, in the key_accounts in
+		// the tx's message to create a new valid threshold signing payload.
+		maybe_current_onchain_key.map_or(RequiresSignatureRefresh::False, |current_on_chain_key| {
+			match call.signer().and_then(|signer| {
+				if signer != current_on_chain_key {
+					Some(signer)
+				} else {
+					None
+				}
+			}) {
+				Some(signer) => {
+					let mut modified_call = (*call).clone();
+					// the unwraps should be safe because we are in the code where on chain key
+					// already exists (see above) and so thecurrent_key_epoch should also exist and
+					// so should the corresponding key,
+					let current_aggkey = SolanaThresholdSigner::keys(
+						SolanaThresholdSigner::current_key_epoch().unwrap(),
+					)
```

Comment:
Is this not equivalent to KeyProvider::active_epoch_key()?
But also, can we not just use the current_on_chain_key?

Reply:
Another thing - ApiCall has a 'refresh_replay_protection' method. We don't seem to use it, which is fine, but we should not leave it `unimplemented`.

Reply:
will change to use active_epoch_key()

We can use ```current_on_chain_key``` i guess but it is a less safe because this apicall will get signed with the current epoch key in the thrshold sig pallet and the key in account_keys has to be the same as the signer so it is safer to use the same storage item to set the key here as will be when the call will be signed.

I was thinking the same about the refresh_replay_protection method but thought maybe do it another PR. I guess I can just do it in this PR.